lib/spack/spack/cmd/common/arguments.py

```diff
@@ -43,6 +53,44 @@ def add_common_arguments(parser, list_of_arguments):
         parser.add_argument(*x.flags, **x.kwargs)
 
 
+def filter_by_tags(specs_or_packages, tags):
+    """Filter the set of specs passed as argument by tags.
+
+    Args:
+        specs_or_packages: list of items to be filtered
+        tags: tags to be used for filtering
+
+    Returns:
+        list of items in specs that have all the tag in tags
+    """
+
+    # If the list is empty, return early
+    if not specs_or_packages:
+        return
+
+    # Else, try to have a uniform access to packages for both list
+    # of specs and list of packages
+    try:
+        # This statement is needed to check if I am dealing
+        # with a list of specs
+        specs_or_packages[0].package
+        package = lambda x: x.package
+    except AttributeError:
+        package = lambda x: x
+
+    # Restrict to elements that have the 'tags' attribute defined
+    has_tags = lambda x: hasattr(package(x), 'tags')
+    specs_or_packages = [x for x in specs_or_packages if has_tags(x)]
+
+    # Filter by required tags
+    for tag in tags:
+        specs_or_packages = [
+            x for x in specs_or_packages if tag in package(x).tags
+        ]
```

Comment:
Concerning the logic here: do you think it should be `any` or `all`? Currently I implemented `all` but was not sure what could be better in case of multiple tags like:
```console
spack find --tags foo --tags bar
```